dev/core/test/com/google/gwt/dev/jjs/impl/JChangeTrackingVisitorTest.java

```diff
@@ -46,7 +48,10 @@ public AddParamWhenEnterMethodVisitor(OptimizerContext optimizerCtx) {
 
     @Override
     public boolean enter(JMethod x, Context ctx) {
-      x.createParameter(SourceOrigin.UNKNOWN, "_newParam_enter", JPrimitiveType.INT);
```

Comment:
How do these changes in this class relate to records?

Reply:
The answer is "Objects.hash(Object...)". Thanks for asking, I had forgotten this headache.

So some parts of gwt-dev depend on gwt-user (while of course gwt-user depends on gwt-dev...) especially for tests. Other tests do not use gwt-user, and mock out their own classes.

The `JavaResourceBase` type is a container of a bunch of partially-implemented classes - just enough to get past the bare bones requirements of "this is essentially Java, and so your code will work. In this patch, we needed to add Record both as Record.java and as a mock. We likewise needed ObjectMethods in both places despite the fact that it has no members - this satisfies JDT, which thinks we're going to need it for runtime MethodHandle creation for each of our Record types. Finally, Objects. 

Objects.hash() is the first method in this super-barebones JRE that uses varargs. 

And JChangeTrackingVisitorTest.AddParamWhenEnterMethodVisitor was implemented assuming there was no such thing as varargs. It is _possible_ (but not very likely) too that JChangeTrackingVisitor itself could have bugs around varargs since it was clearly not tested. Adding Objects.hash() thus requires that this test be updated to allow that varargs are a real thing.

If we'd made a new "mini-JRE" for each test, this wouldn't have happened, since this test wouldn't have failed with the addition of varargs for a different test's JRE. Likewise, if these were all built as integration tests and so used the "real" GWT jre emul, this would have been fixed long ago. But neither of those was done, so I got to find this surprise while implementing the feature.